build/rpmbuild_internal.h

```diff
@@ -31,9 +32,45 @@ struct TriggerFileEntry {
     uint32_t priority;
 };
 
+typedef enum rpmParseLineType_e {
+    LINE_DEFAULT           = 0,
+    LINE_IF                = (1 << 0),
+    LINE_IFARCH            = (1 << 1),
+    LINE_IFNARCH           = (1 << 2),
+    LINE_IFOS              = (1 << 3),
+    LINE_IFNOS             = (1 << 4),
+    LINE_ELSE              = (1 << 5),
+    LINE_ENDIF             = (1 << 6),
+    LINE_INCLUDE           = (1 << 7),
+} rpmParseLineType;
+
+typedef const struct parsedSpecLine_s {
+    int id;
+    size_t textLen;
+    const char *text;
+    int withArgs;
+    int isConditional;
+    int wrongPrecursors;
+    const char *info_text;
+} * parsedSpecLine;
+
+static struct parsedSpecLine_s const lineTypes[] = {
+    { LINE_ENDIF,      LEN_AND_STR("%endif")  , 0, 1, LINE_ENDIF, "with no %%if"},
+    { LINE_ELSE,       LEN_AND_STR("%else")   , 0, 1, LINE_ENDIF | LINE_ELSE, "after %%else" },
+    { LINE_IF,         LEN_AND_STR("%if")     , 1, 1, 0, "after %%if"},
+    { LINE_IFARCH,     LEN_AND_STR("%ifarch") , 1, 1, 0, "after %%ifarch"},
+    { LINE_IFNARCH,    LEN_AND_STR("%ifnarch"), 1, 1, 0, "after %%ifnarch"},
+    { LINE_IFOS,       LEN_AND_STR("%ifos")   , 1, 1, 0, "after %%ifos"},
+    { LINE_IFNOS,      LEN_AND_STR("%ifnos")  , 1, 1, 0, "after %%ifnos"},
```

Comment:
>The %ifs should not have any info text because there are no such error cases, and the texts should be marked for translation.

>Programmatically constructed messages tend to be a problem for translators, might better to have the whole string in one piece, ie "%s with no %s" and "%s after %s" to allow translators to at least reorder the words.

Changed in the new version

>  Also, isn't this now missing the else-with-no-if case, or am I just misreading it?

No in this case it wrote e.g.
/data/SPECS/attrtest.spec:1: %%endif with no %%if
/data/SPECS/configtest.spec:1: %%else with no %%if

> Hmm, "with no" could just as well be changed into "before", which makes the messages more in line with each other: this is just about the order of things, so things are either before or after its proper place.

This message appears e.g. in case that in spec file is only one conditional: %endif. So in this case there is no %if after the %endif, thus changing "with no" into "before" good idea.